src/Bicep.LangServer/Handlers/BicepHoverHandler.cs

```diff
@@ -136,38 +136,38 @@ private static string GetFunctionMarkdown(FunctionSymbol function, FunctionCallS
             var buffer = new StringBuilder();
             buffer.Append($"function ");
             buffer.Append(function.Name);
+
+            if (model.TypeManager.GetMatchedFunctionOverload(functionCall) is { } matchedOverload)
+            {
+                buffer.Append(matchedOverload.TypeSignature);
+                return CodeBlockWithDescription(buffer.ToString(), matchedOverload.Description);
+            }
+
             buffer.Append('(');
```

Comment:
The [hover test] (https://github.com/Azure/bicep/pull/6592/files#diff-48eafb4bca1f825511d491c521677bb7bd60ba514fc360c6f31d51376492965cR422) shows that the `concat` function accept arrays, but that's not true. I wonder if we should just fall back to the first overload instead of constructing a nonexistent signature based on the argument types.

Reply:
The synthetic signature created when no overload is matched is using a union of the types that would be accepted in a given function call position by any of its overloads, which is probably just as confusing as the behavior described in #3640. (With this PR, hovering over `concat({}, null, [], true)` would show a code block of `concat(array|bool|int|string, array|bool|int|string, array|bool|int|string, array|bool|int|string): any` with no description.)

It looks like the hover response returned by the LSP can contain a list of tooltips, which could maybe be used to show all overloads if none is matched?

Reply:
A hover response with a list of tooltips will display them all rather than showing them in a scrollable list:

![hover_no_matching_overload](https://user-images.githubusercontent.com/705500/164262405-7d200c41-4433-4823-978c-fceeaabf839a.png)

I think that hover box is more helpful than a synthetic signature, but the UX might not be ideal if a function had a large number of overloads. I can also add a message to the tooltip, though it does not help with the UX concern:

![hover_no_matching_overload_with_message](https://user-images.githubusercontent.com/705500/164263412-6419290d-6599-4c27-9ea4-8a01e88611b5.png)

It looks like OmniSharp will do what you suggested and display the first overload if none match, so that might be the behavior users are accustomed to:

![hover_no_matching_overload_omnisharp](https://user-images.githubusercontent.com/705500/164264213-aa79f114-091c-4ac2-9b04-074308577bd6.png)

Do you have any preference here?

Reply:
Welcome to the team Jonny!

At first glance, it's kind of nice to see all of the overloads, since we don't know which one is most relevant to you. If we had a reliable way to say "this is the signature that closest matches the arguments that you've provided:" that might be interesting too.

But good to know that this is in contrast to what OmniSharp does. I'd be curious what other languages do for this problem.

Reply:
Pylance shows all potentially matching signatures, which does make it easier to see what your options are. The hover window is resizable and scrollable (at least in VS Code), so an overloaded tooltip is still usable:

![pylance_hover_no_match](https://user-images.githubusercontent.com/705500/164281512-48d2550e-58de-446b-88f3-317dba488eaf.png)

The signature help handler has some logic to determine which overload is the most likely match for a given function call expression; I'll see if I can factor that out and have it shared between the two handlers.

Reply:
TypeScript shows the first overload in VS Code:
![image](https://user-images.githubusercontent.com/16367959/164287162-6c55a0cb-f2a2-4f3f-a915-83709f9c2561.png)

I think showing all potential matching signatures also makes sense.